Add `Prepared For` Constraints

Gabeblis opened the pull request:
# Committer Notes
## Purpose
This PR introduces three constraints aimed at ensuring FedRAMP compliance for the digital authorization package by verifying the `prepared-for` roles/parties. These constraints ensure that all required elements are present and that necessary information is accurately documented.

## Changes
Constraint: `role-defined-prepared-for`
- Validates that there is a defined role with the ID "prepared-for".

Constraint: `responsible-party-prepared-for`
- Ensures that a responsible party is assigned to the prepared-for role.
 
Constraint: `responsible-party-prepared-for-location-valid`
- Validates that the party assigned to the prepared-for role includes a location-uuid with the required address details.

Tests
- Added invalid test data files for all three constraints.
- Added pass/fail YAML files for the three constraints to verify their functionality.
- Updated ssp-all-VALID.xml to contain the required information so that the constraints fire and pass as expected.

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?
- [x] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

aj-stein-gsa reviewed (changes requested):
Can you address the updated AC requirements in the issue? Sorry again.

Gabeblis commented:
> Can you address the updated AC requirements in the issue? Sorry again.

Resolved here: [045e70e](https://github.com/GSA/fedramp-automation/pull/869/commits/045e70ef3055c9c1b1620d783b73690a05747714). I added the check for an embedded address assembly in the party and I added another `-VALID.xml` test file in order to validate both scenarios.

aj-stein-gsa commented on src/validations/constraints/fedramp-external-constraints.xml:
```diff
@@ -235,6 +239,21 @@
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-for-csp"/>
+                <message>A FedRAMP SSP MUST have a responsible party that references a role defined as "prepared-for". </message>
+            </expect>
+            <expect id="responsible-party-prepared-for-location-valid" target="." test="$prepared-for-party/address[@type='work'] and $prepared-for-party/address/addr-line and $prepared-for-party/address/city and $prepared-for-party/address/state and $prepared-for-party/address/postal-code or $prepared-for-location/address[@type='work'] and $prepared-for-location/address/addr-line and $prepared-f ...
+                <formal-name>Responsible Party Prepared For Location Valid</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-for-csp"/>
```
Blocking: I am kind of intrigued. How can this work if the booleans `and`s connect all the expressions and not an `or` if you can have a location UUID, an address inside the party, but not both?

Gabeblis replied:
I'm a little confused by what you're asking. How the expression works is it checks if either a full address is embedded in the party or if it references a location assembly, then check if that has a full address. If both don't exist, or are incomplete, it will return false. If either one exists and is complete, it returns true. This is verified by the invalid test data files and the valid ones. I could make an invalid test file for each scenario, but that would require like 10 extra invalid test files. I manually checked each missing field, and if one missing field causes the expression to return false, then we can assume that more missing fields would also return false. Currently, it allows for both to exist. I could be more specific and say if one exists, then don't allow the other and vice versa.

aj-stein-gsa replied:
Sorry I address above, but you understood perfectly!

aj-stein-gsa replied:
Resolving this one to put the suggestion where I meant it.

aj-stein-gsa commented on src/validations/constraints/fedramp-external-constraints.xml:
```diff
@@ -235,6 +239,21 @@
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-for-csp"/>
+                <message>A FedRAMP SSP MUST have a responsible party that references a role defined as "prepared-for". </message>
+            </expect>
+            <expect id="responsible-party-prepared-for-location-valid" target="." test="$prepared-for-party/address[@type='work'] and $prepared-for-party/address/addr-line and $prepared-for-party/address/city and $prepared-for-party/address/state and $prepared-for-party/address/postal-code or $prepared-for-location/address[@type='work'] and $prepared-for-location/address/addr-line and $prepared-f ...
```
OK you understood perfectly, I just assumed the precedent of `and`/`or` needs some parens. If it doesn't, it seems kind of unreadable to me.

```suggestion
            <expect id="responsible-party-prepared-for-location-valid" target="." test="($prepared-for-party/address[@type='work'] and $prepared-for-party/address/addr-line and $prepared-for-party/address/city and $prepared-for-party/address/state and $prepared-for-party/address/postal-code) or ($prepared-for-location/address[@type='work'] and $prepared-for-location/address/addr-line and $prepared-for-location/address/city and $prepared-for-location/address/state and $prepared-for-location/address/postal-code)" level="WARNING">
```

aj-stein-gsa replied:
I mean your good, I don't the PEMDAS properties of XPath 3.1 or Metapath that well, so I wanted to look into this.

Gabeblis replied:
`and` takes a higher precedence over `or` so both expressions work the same. Definitely easier to read with parenthesis though

Gabeblis replied:
Resolved: [eceab1a](https://github.com/GSA/fedramp-automation/pull/869/commits/eceab1af055db024d11796008301b1c1d84605c0)

kyhu65867 commented on src/validations/constraints/fedramp-external-constraints.xml:
```diff
@@ -176,6 +180,16 @@
+            <expect id="responsible-party-prepared-for-location-valid" target="." test="($prepared-for-party/address[@type='work'] and $prepared-for-party/address/addr-line and $prepared-for-party/address/city and $prepared-for-party/address/state and $prepared-for-party/address/postal-code) or ($prepared-for-location/address[@type='work'] and $prepared-for-location/address/addr-line and $prepare ...
+                <formal-name>Responsible Party Prepared For Location Valid</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-for-csp"/>
+                <message>A FedRAMP SSP MUST have a responsible party for who the document was prepared for, and that party MUST define an address.</message>
```
Since this is a separate constraint, I don't think I would repeat the responsible party prepared for option, and instead have the message just say "A FedRAMP SSP MUST define the address of the responsible party for whom the document was prepared. "

Gabeblis replied:
Resolved: [c67e133](https://github.com/GSA/fedramp-automation/pull/869/commits/c67e1333ab12b1110fe723425468175258eb0ac2)

kyhu65867 commented on src/validations/constraints/fedramp-external-constraints.xml:
```diff
@@ -176,6 +180,16 @@
+            <expect id="responsible-party-prepared-for" target="." test="exists(responsible-party[@role-id eq 'prepared-for'])" level="ERROR">
+                <formal-name>Responsible Party Prepared For</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-for-csp"/>
+                <message>A FedRAMP SSP MUST have a responsible party that defines the party that the document was prepared for.</message>
```
Change message slightly to "A FedRAMP SSP MUST have a responsible party that defines the party for whom the document was prepared." (better grammar)

aj-stein-gsa replied:
I do not disagree, I just felt the more formal, rigid grammatical style feels a little off from the more approachable, plain language style. I agree it is more casual and repetitive with the redundant "thats," but I defer to team opinion here, not just AJ.

Gabeblis replied:
I don't mind the change. Resolved: Resolved: [c67e133](https://github.com/GSA/fedramp-automation/pull/869/commits/c67e1333ab12b1110fe723425468175258eb0ac2)

kyhu65867 reviewed (dismissed):
tweaks to message clarity

aj-stein-gsa commented:
@Gabeblis sorry to be that person, but can you rebase?
